app/vmalert/group.go

```diff
@@ -238,10 +265,15 @@ func (e *executor) execConcurrently(ctx context.Context, rules []Rule, concurren
 		returnSeries = true
 	}
 
-	if concurrency == 1 {
-		// fast path
+	if concurrency == 1 { // fast path
 		for _, rule := range rules {
-			res <- e.exec(ctx, rule, returnSeries, interval)
+			select {
+			// break rules eval if context was cancelled
+			case <-ctx.Done():
+				break
```

Comment:
Hm, interesting. I didn't take into account such case with very frequent reloads. What I want to achieve here is to improve `vmalert` user experience in this case and try to keep behaviour similar to Prometheus one, since people get used to it. 
In your case, would it help if groups will start to update concurrently and then sync before finish? So the config reload time in the worst case would take the time of the slowest group in configuration.

Reply:
it will help.  
But still, we have to wait for current reloading finished before we can start another reloading. In the worst case(which might be common), one long time cost single group will block other reloading requests.

Reply:
I think the task must be attacked from another angle: new and old rules must be compared during config reload and then only the modified rules must be stopped and updated, while unmodified rules must continue executing without stopping.

Reply:
Looks good to me @valyala . 
And "update groups concurrently" is still needed here.

Reply:
> and old rules must be compared during config reload and then only the modified rules must be stopped and updated

Do you mean `groups`? Rules are executed sequentially and belong to the group. And group entity is a unit of work that needs to be executed. Evaluation period defines the frequency for the group to evaluate its rules and all groups are running concurrently. So when we update group - vmalert stops the execution to update the rules and then continues execution from beginning. The improvement I see here is that we can skip `group` update if its rules didn't change.

Reply:
If stop and update the entire group, some rules(under the canceled group) may get canceled together. If the reload is frequent(which we can not control) and the group is large, some rules may never get evaluated.  
But if the cancel unit is rule itself, we can interrupt the specific rule if it was modified and some other not updated rules can still run its' evaluation.

Reply:
About rule evaluation, plz take a look at PR #724 .  
This PR balance the group evaluation into different pieces of interval. It should reduce the load of storage nodes.

Reply:
>  So when we update group - vmalert stops the execution to update the rules and then continues execution from beginning. The improvement I see here is that we can skip group update if its rules didn't change.

Looks good to me. I'd propose starting with this task and then later thinking about per-rule granularity during updates.

Reply:
@valyala @dxtrzhang plz see https://github.com/VictoriaMetrics/VictoriaMetrics/pull/759